File: lib/rayz/chapter7.rb
require_relative "world"
require_relative "camera"
require_relative "point_light"
require_relative "point"
require_relative "color"
require_relative "sphere"
require_relative "transformations"
require_relative "material"

module Rayz
  class Chapter7
    def self.run
      puts "Chapter 7: Making a Scene"

      # Create a world
      world = World.new

      # Add light source
      world.light = PointLight.new(
        Point.new(x: -10, y: 10, z: -10),
        Color.new(red: 1, green: 1, blue: 1)
      )

      # Create floor sphere (scaled and flattened)
      floor = Sphere.new
      floor.transform = Transformations.scaling(10, 0.01, 10)
      floor.material.color = Color.new(red: 1, green: 0.9, blue: 0.9)
      floor.material.specular = 0
      world.objects << floor

      # Create left wall
      left_wall = Sphere.new
      left_wall.transform = Transformations.translation(0, 0, 5) *
        Transformations.rotation_y(-Math::PI / 4) *
        Transformations.rotation_x(Math::PI / 2) *
        Transformations.scaling(10, 0.01, 10)
      left_wall.material = floor.material
      world.objects << left_wall

      # Create right wall
      right_wall = Sphere.new
      right_wall.transform = Transformations.translation(0, 0, 5) *
        Transformations.rotation_y(Math::PI / 4) *
        Transformations.rotation_x(Math::PI / 2) *
        Transformations.scaling(10, 0.01, 10)
      right_wall.material = floor.material
      world.objects << right_wall

      # Create middle sphere
      middle = Sphere.new
      middle.transform = Transformations.translation(-0.5, 1, 0.5)
      middle.material.color = Color.new(red: 0.1, green: 1, blue: 0.5)
      middle.material.diffuse = 0.7
      middle.material.specular = 0.3
      world.objects << middle

      # Create right sphere
      right = Sphere.new
      right.transform = Transformations.translation(1.5, 0.5, -0.5) *
        Transformations.scaling(0.5, 0.5, 0.5)
      right.material.color = Color.new(red: 0.5, green: 1, blue: 0.1)
      right.material.diffuse = 0.7
      right.material.specular = 0.3
      world.objects << right

      # Create left sphere
      left = Sphere.new
      left.transform = Transformations.translation(-1.5, 0.33, -0.75) *
        Transformations.scaling(0.33, 0.33, 0.33)
      left.material.color = Color.new(red: 1, green: 0.8, blue: 0.1)
      left.material.diffuse = 0.7
      left.material.specular = 0.3
      world.objects << left

      # Set up camera
      camera = Camera.new(400, 200, Math::PI / 3)
      camera.transform = Transformations.view_transform(
        Point.new(x: 0, y: 1.5, z: -5),
        Point.new(x: 0, y: 1, z: 0),
        Vector.new(x: 0, y: 1, z: 0)
      )

      puts "Rendering scene (400x200 pixels)..."
      puts "This may take a minute..."

      # Render the scene
      canvas = camera.render(world)

      # Write to file
      File.write("chapter7.ppm", canvas.to_ppm)
      puts "Scene rendered to chapter7.ppm"
      puts "\n" + ("=" * 60) + "\n"
    end
  end
end

File: python/examples/chapter7.py
"""Chapter 7: Making a Scene — world, camera, and shadows."""

from __future__ import annotations

import math
import os

from rayz.camera import Camera
from rayz.color import Color
from rayz.point_light import PointLight
from rayz.sphere import Sphere
from rayz.transformations import (
    rotation_x,
    rotation_y,
    scaling,
    translation,
    view_transform,
)
from rayz.tuple import Point, Vector
from rayz.world import World


def run() -> None:
    print("\n=== Chapter 7: Making a Scene ===\n")

    world = World()
    world.light = PointLight(Point(-10, 10, -10), Color(1, 1, 1))

    floor = Sphere()
    floor.set_transform(scaling(10, 0.01, 10))
    floor.material.color = Color(1, 0.9, 0.9)
    floor.material.specular = 0
    world.objects.append(floor)

    left_wall = Sphere()
    left_wall.set_transform(
        translation(0, 0, 5) * rotation_y(-math.pi / 4) * rotation_x(math.pi / 2) * scaling(10, 0.01, 10)
    )
    left_wall.material = floor.material
    world.objects.append(left_wall)

    right_wall = Sphere()
    right_wall.set_transform(
        translation(0, 0, 5) * rotation_y(math.pi / 4) * rotation_x(math.pi / 2) * scaling(10, 0.01, 10)
    )
    right_wall.material = floor.material
    world.objects.append(right_wall)

    middle = Sphere()
    middle.set_transform(translation(-0.5, 1, 0.5))
    middle.material.color = Color(0.1, 1, 0.5)
    middle.material.diffuse = 0.7
    middle.material.specular = 0.3
    world.objects.append(middle)

    right = Sphere()
    right.set_transform(translation(1.5, 0.5, -0.5) * scaling(0.5, 0.5, 0.5))
    right.material.color = Color(0.5, 1, 0.1)
    right.material.diffuse = 0.7
    right.material.specular = 0.3
    world.objects.append(right)

    left = Sphere()
    left.set_transform(translation(-1.5, 0.33, -0.75) * scaling(0.33, 0.33, 0.33))
    left.material.color = Color(1, 0.8, 0.1)
    left.material.diffuse = 0.7
    left.material.specular = 0.3
    world.objects.append(left)

    camera = Camera(200, 100, math.pi / 3)
    camera.transform = view_transform(Point(0, 1.5, -5), Point(0, 1, 0), Vector(0, 1, 0))

    print("Rendering 200x100 scene with shadows...")
    canvas = camera.render_parallel(world)

    out_path = os.path.join(os.path.dirname(__file__), "chapter7.ppm")
    print(f"Writing {out_path}...", end="", flush=True)
    with open(out_path, "w") as f:
        f.write(canvas.to_ppm())
    print(" done.")
    print("Scene with floor, walls, and three spheres rendered.")
    print("\n" + "=" * 60 + "\n")


if __name__ == "__main__":
    run()
